substrate-node/pallets/pallet-smart-contract/src/lib.rs

```diff
@@ -2256,6 +2268,42 @@ impl<T: Config> Pallet<T> {
 
         Ok(().into())
     }
+
+    pub fn _attach_solution_provider_id(
+        account_id: T::AccountId,
+        contract_id: u64,
+        solution_provider_id: u64,
+    ) -> DispatchResultWithPostInfo {
+        ensure!(
+            SolutionProviders::<T>::contains_key(solution_provider_id),
+            Error::<T>::NoSuchSolutionProvider
+        );
+
+        let mut contract = Contracts::<T>::get(contract_id).ok_or(Error::<T>::ContractNotExists)?;
+
+        let twin_id = pallet_tfgrid::TwinIdByAccountID::<T>::get(&account_id)
+            .ok_or(Error::<T>::TwinNotExists)?;
+
+        ensure!(
+            contract.twin_id == twin_id,
+            Error::<T>::UnauthorizedToChangeSolutionProviderId
+        );
+
+        match contract.solution_provider_id {
+            Some(_) => {
```

Comment:
By doing this we allow to attach a solution provider id to contract only if there is not an existing one
What if we want to change the provider id?

Reply:
Yes indeed